### PR TITLE
[ao] Adding error message invalid conv(T) padding in quantization

### PR DESCRIPTION
**Pull request by HDCharles**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #76475

Summary: per https://github.com/pytorch/pytorch/issues/76304 it's
confusing for the quantization flow to throw a generic C++ RuntimeError
when a module with a valid fp32 conv argument is quantized so a clearer
error has been added

Test Plan: python test/test_quantization.py
TestStaticQuantizedModule.test_quantized_conv_padding_error

python test/test_Quantization.py
TestStaticQuantizedModule.test_quantized_conv_transpose_padding_error

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35986965](https://our.internmc.facebook.com/intern/diff/D35986965)

### CODE REVIEW

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35986965).

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35986965).

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35986965).

**Comment by HDCharles** [comment]
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35986965).

File: torch/nn/quantized/modules/conv.py
Comment (z-a-f) [inline torch/nn/quantized/modules/conv.py:58]: I think the padding is supposed to be int, tuple, or str? At least according to the nn.ConvXd docs: https://pytorch.org/docs/stable/generated/torch.nn.Conv1d.html

**Review by z-a-f — CHANGES_REQUESTED** [review]
I think this will always throw an error: we use `_single`, `_pair`, etc. to convert the padding arguments to tuples.

File: torch/nn/quantized/modules/conv.py
Comment (HDCharles) [reply torch/nn/quantized/modules/conv.py:58]: The problem is the quantized kernel doesn't actually handle them, it only accepts a list[int] which is what a tuple ad this point gets converted to. See the github issue this PR is in reference to. note: We're looking for tuple(str) because for conv 2d or whatever, when they do _single or _pair on the string, that's the result.

**Comment by z-a-f** [comment]
> I think this will always throw an error: we use `_single`, `_pair`, etc. to convert the padding arguments to tuples.

I take it back -- that conversion happens after this line. My bad.

File: torch/nn/quantized/modules/conv.py
Comment (z-a-f) [reply torch/nn/quantized/modules/conv.py:58]: I don't have a super-strong opinion about when the string arguments should be implemented, but I do believe that quant API should be the same as the FP. Ipso facto, I would rather see the padding handling be fixed, rather than telling the users that we don't support that.

---

> The problem is the quantized kernel doesn't actually handle them,

The kernels don't really have to make sure the arguments are not malformed. On the python side, we can just add all the necessary plumbing, without being blocked by the kernel dev.

With that said, I agree that the FP kernels do have support for string padding. In FP the conv kernels are overloaded, and there are two implementations:
1. [with type of padding being `IntArrayRef`](https://github.com/pytorch/pytorch/blob/ba818991afb4ef6189fa4ed57dd204173e9c7b74/aten/src/ATen/native/Convolution.cpp#L764-L781) -- this is just taking the padding as an array of integers.
2. [with type of padding being `std::string`](https://github.com/pytorch/pytorch/blob/ba818991afb4ef6189fa4ed57dd204173e9c7b74/aten/src/ATen/native/Convolution.cpp#L906-L920) -- this does the [same conversion](https://github.com/pytorch/pytorch/blob/ba818991afb4ef6189fa4ed57dd204173e9c7b74/aten/src/ATen/native/Convolution.cpp#L847-L857) as [in python](https://github.com/pytorch/pytorch/blob/ba818991afb4ef6189fa4ed57dd204173e9c7b74/torch/nn/modules/conv.py#L114-L123), and calls the convolution described in (1)

I think the reason there are two versions of the convXd is because one can call `F.conv1d(padding='same')`, and it is necessary by API to support that. It is not super easy to implement the conversion directly in the `nn/functional.py`, as it [uses an `_add_docstr` wrapper](https://github.com/pytorch/pytorch/blob/ba818991afb4ef6189fa4ed57dd204173e9c7b74/torch/nn/functional.py#L30), instead of direct function definition.

However, for the quantized we don't have to do it that way:

1. In the `nn.quantized.ConvXd` we can add a routine to convert the padding
    1. 'valid': `self.padding=0`
    2. 'same': `self.padding=[self.kernel_size//2, self.kernel_size//2]` (in practice, we need to handle the dilation and different dimensionality).
1. In the [`nn.quantized.functional.convXd`](https://github.com/pytorch/pytorch/blob/ba818991afb4ef6189fa4ed57dd204173e9c7b74/torch/nn/quantized/functional.py#L111-L168) we can add the same routine to convert the padding

**Review by z-a-f — APPROVED** [review]
Accepting to unblock -- will work on implementing the 'same' and 'valid' padding modes

**Pull request closed without merging**
